app/models/user.rb

```diff
@@ -107,6 +93,43 @@ def decorate
     UserDecorator.new(self)
   end
 
+  def max_login_attempts?
+    second_factor_attempts_count.to_i >= max_login_attempts
+  end
+
+  def max_login_attempts
+    3
+  end
+
+  def create_direct_otp
+    update(
+      direct_otp: random_base10(TwoFactorAuthenticatable.direct_otp_length),
+      direct_otp_sent_at: Time.now.utc,
+    )
+  end
+
+  def generate_totp_secret
+    ROTP::Base32.random_base32
+  end
+
+  def authenticate_direct_otp(code)
+    return false if direct_otp.nil? || direct_otp != code || direct_otp_expired?
```

Comment:
I wasn't sure if the `!=` should be a `secure_compare`?

Reply:
IMO comparison is not susceptible to timing attack since are not related to the crypto-signaturesand not an API that an attacker can hit and guess

Reply:
Thanks yeah, it doesn't seem strictly necessary given the rate limiting that's already in place 🙂